### PR TITLE
make_graph update versions from .json

### PR DESCRIPTION
**Pull request by viniciusdc**
Continuing the work from [separate version updates](https://github.com/regro/cf-scripts/issues/842) and [issue](https://github.com/regro/cf-scripts/pull/1049).

### CODE REVIEW

File: conda_forge_tick/make_graph.py [hunk 1]
```diff
@@ -352,6 +352,25 @@ def update_nodes_with_bot_rerun(gx):
...
+    """Updates every node with it's new version (when available)"""
+    # check if the versions folder is available
+    if os.path.isdir("./versions"):
+        # get all the available node.json files
+        pass
+    else:
+        return
+    list_files = os.listdir("./versions/")
+    for file in list_files:
+        node = str(file).replace(".json", "")
```
Comment (CJ-Wright) [inline conda_forge_tick/make_graph.py:365]: why are you removing `.json`?

File: conda_forge_tick/make_graph.py [hunk 2]
```diff
@@ -352,6 +352,25 @@ def update_nodes_with_bot_rerun(gx):
...
+        with open(f"./versions/{file}") as json_file:
+            version_data = json.load(json_file)
+        with gx.nodes[f"{node}"]["payload"] as attrs:
+            attrs["new_version"] = version_data.get("new_version")
```
Comment (CJ-Wright) [inline conda_forge_tick/make_graph.py:369]: `attrs.update(**version_data))`

Comment (CJ-Wright) [reply conda_forge_tick/make_graph.py:369]: If that doesn't work please add an update method to `LazyJSON`

File: conda_forge_tick/make_graph.py (hunk 1 above)
Comment (beckermr) [reply conda_forge_tick/make_graph.py:365]: To compute the node name in the graph I think

Comment (viniciusdc) [reply conda_forge_tick/make_graph.py:365]: To get the node name to update the attrs information regarding that node

File: conda_forge_tick/make_graph.py (hunk 2 above)
Comment (viniciusdc) [reply conda_forge_tick/make_graph.py:369]: I will try that, but I will have to think a way to test that though

Comment (viniciusdc) [reply conda_forge_tick/make_graph.py:369]: Should I add the bad feedstocks too for `update_versions` ? to solve the problems with the outdated `.json` files ?

Comment (viniciusdc) [reply conda_forge_tick/make_graph.py:369]: Or let it run for now and see that after ?

Comment (viniciusdc) [reply conda_forge_tick/make_graph.py:369]: I was looking into the `LazyJSON` config at `.utils` and there isn't any `.update` method there, so would this be enough ?
```python
    def __update__(self, data: dict) -> None:
        for key in data.items():
            self.__dict__[f'{key}'] = data[f'{key}']
```

File: conda_forge_tick/utils.py [hunk 3]
```diff
@@ -182,6 +180,11 @@ def __enter__(self) -> "LazyJson":
     def __exit__(self, *args: Any) -> Any:
         self._dump()
 
+    def __update__(self, data: dict) -> None:
```
Comment (CJ-Wright) [inline conda_forge_tick/utils.py:183]: I don't think this needs to be `__update__` it should be `update`

File: conda_forge_tick/utils.py [hunk 4]
```diff
@@ -182,6 +180,11 @@ def __enter__(self) -> "LazyJson":
...
+        self._load()
+        for key in data.items():
+            self.data[f"{key}"] = data[f"{key}"]
```
Comment (CJ-Wright) [inline conda_forge_tick/utils.py:186]: I would use the object's `setitem`: `self[key] = data[key]` that way you don't need to do the load and dump

File: conda_forge_tick/utils.py (hunk 3 above)
Comment (viniciusdc) [reply conda_forge_tick/utils.py:183]: I will change that thanks :+1:

File: conda_forge_tick/utils.py (hunk 4 above)
Comment (viniciusdc) [reply conda_forge_tick/utils.py:186]: uhm, you are right

Comment (CJ-Wright) [reply conda_forge_tick/utils.py:186]: I would just use the `[]` syntax directly. Also you don't need an f-string for the keys, they are already strings.

Comment (viniciusdc) [reply conda_forge_tick/utils.py:186]: > I would just use the `[]` syntax directly. Also you don't need an f-string for the keys, they are already strings.

Oh thanks, I was not sure about the syntax usage.

File: conda_forge_tick/utils.py [hunk 5]
```diff
@@ -180,10 +180,9 @@ def __enter__(self) -> "LazyJson":
     def __exit__(self, *args: Any) -> Any:
         self._dump()
 
-    def __update__(self, data: dict) -> None:
-        self._load()
-        for key in data.items():
-            self.data[f"{key}"] = data[f"{key}"]
+    def _update(self, data: dict) -> None:
```
Comment (beckermr) [inline conda_forge_tick/utils.py:183]: why did you rename this?

Comment (viniciusdc) [reply conda_forge_tick/utils.py:183]: It was an idea from CJ: "I don't think this needs to be `__update__` it should be `update`"

Comment (beckermr) [reply conda_forge_tick/utils.py:183]: you have an extra underscore at the front.

Comment (viniciusdc) [reply conda_forge_tick/utils.py:183]: Yup, I saw with your comment thanks 👍

Comment (viniciusdc) [reply conda_forge_tick/utils.py:183]: Uhm @beckermr I am getting a problem with this statement `update(self, data: dict)` it's saying that it doesn't match the method base definition of a `MutableMapping` wich should be something as `update(self, data: dict, **kwargs)`. Do you know what can I do to solve this ?

Comment (CJ-Wright) [reply conda_forge_tick/utils.py]: You could structure this as:
```python
def update(self, data, **kwargs):
    self._load()
    self._data.update(data, **kwargs)
    self._dump()
```

Comment (viniciusdc) [reply conda_forge_tick/utils.py]: Ok, Thank you very much 👍

**Comment by CJ-Wright** [comment]
It seems that the tests aren't passing

**Comment by viniciusdc** [comment]
> It seems that the tests aren't passing

Yes, the `tests/test_audit.py::test_grayskull_audit_feedstock` failed

**Comment by CJ-Wright** [comment]
@viniciusdc I think that is due to the new grayskull release

**Comment by CJ-Wright** [comment]
This has conflicts now (although I think I fixed the grayskull issue, at least for the moment)

**Comment by viniciusdc** [comment]
> This has conflicts now (although I think I fixed the grayskull issue, at least for the moment)

I will try to solve the conflicts in a pull, if it's not enough I will create a new PR with this changes.

**Comment by viniciusdc** [comment]
Sorry for the delay, not enough time this weekend. I will try to solve this ASAP.

**Comment by viniciusdc** [comment]
Solved! 😄 (I think)

**Comment by CJ-Wright** [comment]
It seems the build is failing

**Comment by viniciusdc** [comment]
> It seems the build is failing

It is the same error as before `test_grayskull_audit_feedstock` but I don't know what is wrong, there are some instances as:
```python
/tmp/grayskull-depfinder-zbc3slmg/depfinder-2.3.0/versioneer.py:969: DeprecationWarning: invalid escape sequence \s
    '''
```

**Comment by CJ-Wright** [comment]
Ahh, can you merge with master and fix the grayskull test? (There's a commented out section that you'll want to uncomment and remove the currently uncommented seciton)

**Comment by viniciusdc** [comment]
> Ahh, can you merge with master and fix the grayskull test? (There's a commented out section that you'll want to uncomment and remove the currently uncommented seciton)

Yes I can do it, but where is Grayskull test files ? and the section I need to uncomment and merge is easy to find ? (I am asking this not for easiness, is just to do be sure I will do the right thing)

**Comment by CJ-Wright** [comment]
yes that's the right file. The comments are right at the top

**Comment by viniciusdc** [comment]
> yes that's the right file. The comments are right at the top

Is this regro/cf-scripts#1056 ?

**Comment by viniciusdc** [comment]
Really we are gonna test it ? Oh it's so interesting to see if it works 😃
